File: reference/jira-lite.py
#!/usr/bin/env python3
"""
jira-lite CLI - REFERENCE ANSWER for the trainer.

This is what the participants should end up with: the SAME logic as server.py,
exposed as a plain CLI the agent calls on demand instead of a standing MCP socket.

Usage:
    python jira-lite.py list
    python jira-lite.py list --assignee me
    python jira-lite.py show PROJ-101

No SDK needed - just the standard library. That is part of the lesson: a CLI
front-end is lighter than an MCP server both to write and to run.
"""
import argparse
import json
import pathlib

# note: reference/ sits one level below the data, so go up to find fake-jira.json
DATA = json.loads(
    (pathlib.Path(__file__).parent.parent / "fake-jira.json").read_text(encoding="utf-8")
)


def list_issues(assignee=None):
    rows = DATA["issues"]
    if assignee:
        rows = [i for i in rows if i["assignee"] == assignee]
    return [{"key": i["key"], "status": i["status"], "summary": i["summary"]} for i in rows]


def get_issue(key):
    for i in DATA["issues"]:
        if i["key"] == key:
            return i
    return {"error": f"no issue {key}"}


def main():
    p = argparse.ArgumentParser(prog="jira-lite")
    sub = p.add_subparsers(dest="cmd", required=True)

    pl = sub.add_parser("list", help="list issues")
    pl.add_argument("--assignee")

    ps = sub.add_parser("show", help="show one issue by key")
    ps.add_argument("key")

    a = p.parse_args()
    if a.cmd == "list":
        print(json.dumps(list_issues(a.assignee), indent=2))
    elif a.cmd == "show":
        print(json.dumps(get_issue(a.key), indent=2))


if __name__ == "__main__":
    main()

File: reference/jira-lite.go
// jira-lite CLI in Go - REFERENCE ANSWER for the trainer (Go track).
//
// Same logic as go/server.go, exposed as a plain CLI the agent calls on demand
// instead of a standing MCP socket. Standard library only - no MCP SDK.
//
// Run (single file, no module needed - stdlib only):
//
//	go run reference/jira-lite.go list
//	go run reference/jira-lite.go list me
//	go run reference/jira-lite.go show PROJ-101
package main

import (
	"encoding/json"
	"fmt"
	"os"
)

type Issue struct {
	Key         string `json:"key"`
	Status      string `json:"status"`
	Assignee    string `json:"assignee"`
	Summary     string `json:"summary"`
	Updated     string `json:"updated"`
	Description string `json:"description"`
}

type dataset struct {
	Issues []Issue `json:"issues"`
}

func loadIssues() []Issue {
	for _, path := range []string{"fake-jira.json", "../fake-jira.json"} {
		b, err := os.ReadFile(path)
		if err != nil {
			continue
		}
		var d dataset
		if err := json.Unmarshal(b, &d); err != nil {
			fmt.Fprintln(os.Stderr, "parse error:", err)
			os.Exit(1)
		}
		return d.Issues
	}
	fmt.Fprintln(os.Stderr, "fake-jira.json not found (run from the demo-kit-mcp folder)")
	os.Exit(1)
	return nil
}

func printJSON(v any) {
	b, _ := json.MarshalIndent(v, "", "  ")
	fmt.Println(string(b))
}

func main() {
	if len(os.Args) < 2 {
		fmt.Fprintln(os.Stderr, "usage: jira-lite list [assignee] | show <key>")
		os.Exit(2)
	}
	issues := loadIssues()

	switch os.Args[1] {
	case "list":
		assignee := ""
		if len(os.Args) > 2 {
			assignee = os.Args[2]
		}
		type row struct {
			Key     string `json:"key"`
			Status  string `json:"status"`
			Summary string `json:"summary"`
		}
		rows := []row{}
		for _, i := range issues {
			if assignee != "" && i.Assignee != assignee {
				continue
			}
			rows = append(rows, row{i.Key, i.Status, i.Summary})
		}
		printJSON(rows)

	case "show":
		if len(os.Args) < 3 {
			fmt.Fprintln(os.Stderr, "usage: jira-lite show <key>")
			os.Exit(2)
		}
		for _, i := range issues {
			if i.Key == os.Args[2] {
				printJSON(i)
				return
			}
		}
		printJSON(map[string]string{"error": "no issue " + os.Args[2]})

	default:
		fmt.Fprintln(os.Stderr, "unknown command:", os.Args[1])
		os.Exit(2)
	}
}
